sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala

```diff
@@ -775,6 +777,32 @@ private[sql] object ParquetRelation2 extends Logging {
     })
   }
 
+  /**
+   * Returns the original schema from the Parquet file with any missing nullable fields from the
+   * Hive Metastore schema merged in.
+   *
+   * When constructing a DataFrame from a collection of structured data, the resulting object has
+   * a schema corresponding to the union of the fields present in each element of the collection.
+   * Spark SQL simply assigns a null value to any field that isn't present for a particular row.
+   * In some cases, it is possible that a given table partition stored as a Parquet file doesn't
+   * contain a particular nullable field in its schema despite that field being present in the
+   * table schema obtained from the Hive Metastore. This method returns a schema representing the
+   * Parquet file schema along with any additional nullable fields from the Metastore schema
+   * merged in.
+   */
+  private[parquet] def mergeMissingNullableFields(
+      metastoreSchema: StructType,
+      parquetSchema: StructType): StructType = {
+    val fieldMap = metastoreSchema.map(f => f.name.toLowerCase -> f).toMap
+    val missingFields = metastoreSchema
+      .map(_.name.toLowerCase)
+      .diff(parquetSchema.map(_.name.toLowerCase))
+      .map(fieldMap(_))
+      .filter(_.nullable)
+    StructType(parquetSchema ++ missingFields)
```

Comment:
I see. Based on the change made in #5141, it looks like the schema returned by **mergeMissingNullableFields()** will still contain any additional fields defined in `parquetSchema` (lines 766-767). How would you feel about simply removing the additional `parquetSchema` fields in the **mergeMissingNullableFields()** method?

Execution would look something like this:

* remove additional `parquetSchema` fields
* call **mergeMissingNullableFields()** on schema w/additional fields removed
* proceed with executing **mergeMetastoreParquetSchema()** with additions made #5141 removed (should be unnecessary if we prune additional fields first)